File: MaxMinCul.py
#방법1
N = int(input())
nums = list(map(int, input().split()))
add, mns, mul, div = map(int, input().split())

minR = 1e9
maxR = -1e9

def Solution(i, j, add, mns, mul, div):
    global maxR, minR
    if i == N:
        maxR = max(j, maxR)
        minR = min(j, minR)
        return

    else:
        if add:
            Solution(i+1, j+nums[i], add-1, mns, mul, div)
        if mns:
            Solution(i+1, j-nums[i], add, mns-1, mul[2], div[3])
        if mul:
            Solution(i+1, j*nums[i], add, mns, mul-1, div)
        if div:
            Solution(i+1, int(j/nums[i]), add, mns, mul, div-1)

Solution(1, nums[0], add, mns, mul, div)
print(maxR)
print(minR)


#방법2
n = int(input())
arr = list(map(int,input().split()))
add, sub, mul, div = map(int,input().split())

minValue = 1e9
maxValue =-1e9

def dfs(i, now):
  global minValue, maxValue,add,sub,mul,div
  if i == n:
    minValue = min(minValue,now)
    maxValue = max(maxValue,now)
    return
  else:
    if add:
      add -= 1
      dfs(i+1, now+arr[i])
      add += 1
    if sub:
      sub -= 1
      dfs(i+1, now-arr[i])
      sub += 1
    if mul:
      mul -= 1
      dfs(i+1, now*arr[i])
      mul += 1
    if div:
      div -= 1
      dfs(i+1, now//arr[i])
      div += 1

dfs(1,arr[0])
print(maxValue)
print(minValue)

File: MaxMinCul.java
import java.util.*;

class Main {
  public static int minValue = 1000000000;
  public static int maxValue = -1000000000;
  public static int n, add, sub, mul, div;
  public static ArrayList<Integer> arr = new ArrayList<>();

  public static void dfs(int i, int now){
    if (i==n) {
      minValue = Math.min(minValue, now);
      maxValue = Math.max(maxValue, now);
    }
    else{
      if (add > 0){
        add -= 1;
        dfs(i + 1, now + arr.get(i));
        add += 1;
      }
      if (sub > 0){
        sub -= 1;
        dfs(i + 1, now - arr.get(i));
        sub += 1;
      }
      if (mul > 0){
        mul -= 1;
        dfs(i + 1, now * arr.get(i));
        mul += 1;
      }
      if (div > 0){
        div -= 1;
        dfs(i + 1, now / arr.get(i));
        div += 1;
      }
    }
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    n = sc.nextInt();
    for (int i = 0; i < n; i++) {
      int x = sc.nextInt();
      arr.add(x);
    }
    add = sc.nextInt();
    sub = sc.nextInt();
    mul = sc.nextInt();
    div = sc.nextInt();

    dfs(1,arr.get(0));
    System.out.println(maxValue);
    System.out.println(minValue);

  }
}
  
